alchemist/alchemist-fxui/src/main/kotlin/it/unibo/alchemist/boundary/interactions/SelectionHelper.kt

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) 2010-2020, Danilo Pianini and contributors
+ * listed in the main project's alchemist/build.gradle.kts file.
+ *
+ * This file is part of Alchemist, and is distributed under the terms of the
+ * GNU General Public License, with a linking exception,
+ * as described in the file LICENSE in the Alchemist distribution's top directory.
+ */
+
+package it.unibo.alchemist.boundary.interactions
+
+import it.unibo.alchemist.boundary.intersectingNodes
+import it.unibo.alchemist.boundary.makePoint
+import it.unibo.alchemist.boundary.makeRectangleWith
+import it.unibo.alchemist.boundary.wormhole.interfaces.Wormhole2D
+import it.unibo.alchemist.model.interfaces.Node
+import it.unibo.alchemist.model.interfaces.Position2D
+import java.awt.Point
+
+/**
+ * Manages multi-element selection and click-selection.
+ */
+class SelectionHelper<T, P : Position2D<P>> {
+
+    /**
+     * Allows basic multi-element box selections.
+     * @param anchorPoint the starting and unchanging [Point] of the selection
+     */
+    class SelectionBox(val anchorPoint: Point, private val movingPoint: Point = anchorPoint) {
+        /**
+         * The rectangle representing the box.
+         * If the rectangle's dimensions are (0, 0), the rectangle is to be considered non-existing.
+         */
+        val rectangle
+            get() = anchorPoint.makeRectangleWith(movingPoint)
+
+        override fun toString(): String = "[$anchorPoint, $movingPoint]"
+    }
+
+    private var box: SelectionBox? = null
+    private var selectionPoint: Point? = null
+    private var isSelecting = false
+
+    /**
+     * The rectangle representing the box.
+     * If the rectangle's dimensions are (0, 0), the rectangle is to be considered non-existing.
+     */
+    val rectangle
+        get() = box?.rectangle ?: makePoint(0, 0).let { it.makeRectangleWith(it) }
+
+    /**
+     * Begins a new selection at the given point.
+     */
+    fun begin(point: Point): SelectionHelper<T, P> = apply {
+        isSelecting = true
+        selectionPoint = point
+        box = SelectionBox(point)
+    }
+
+    /**
+     * Updates the selection with a new point.
+     */
+    fun update(point: Point): SelectionHelper<T, P> = apply {
+        if (isSelecting) {
+            box?.let {
+                box = SelectionBox(
+                    it.anchorPoint,
+                    point
+                )
```

Comment:
This is extremely unclear. Is this a:
```kotlin
if (isSelecting && box != null) {
    box = SelectionBox(box.anchor, point)
}
```
or am I wrong?

Reply:
The smart cast does not work here, since box is mutable. It would work with the `!!` operator though, what should I do?
It's a bit messy regardless, I could look into refactoring this class in a way that we don't need null checks.

Reply:
Try to please